File: internal/notify/slack.go
package notify

import (
	"context"
	"net/http"
	"strings"
)

// slackNotifier posts to a Slack incoming webhook. The minimal accepted payload
// is {"text": "..."}.
type slackNotifier struct {
	url   string
	httpc *http.Client
}

// NewSlack builds a Slack notifier for the given incoming-webhook URL.
func NewSlack(url string) Notifier {
	return &slackNotifier{url: url, httpc: defaultClient()}
}

func (s *slackNotifier) Notify(ctx context.Context, m Message) error {
	return postJSON(ctx, s.httpc, s.url, map[string]string{"text": slackText(m)})
}

// slackText renders a Message as Slack mrkdwn.
func slackText(m Message) string {
	var b strings.Builder
	if m.Title != "" {
		b.WriteString("*")
		b.WriteString(m.Title)
		b.WriteString("*")
	}
	if m.Text != "" {
		if b.Len() > 0 {
			b.WriteString("\n")
		}
		b.WriteString(m.Text)
	}
	if m.Link != "" {
		if b.Len() > 0 {
			b.WriteString("\n")
		}
		b.WriteString("<")
		b.WriteString(m.Link)
		b.WriteString(">")
	}
	return b.String()
}

File: kotlin/src/main/kotlin/com/automation/agent/notify/Slack.kt
package com.automation.agent.notify

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import java.net.http.HttpClient

/**
 * Posts to a Slack incoming webhook. The minimal accepted payload is {"text": "..."}.
 */
internal class SlackNotifier(
    private val url: String,
    private val client: HttpClient = defaultClient(),
) : Notifier {
    override suspend fun notify(message: Message) {
        postJson(client, url, buildJsonObject { put("text", slackText(message)) })
    }
}

/** Builds a Slack notifier for the given incoming-webhook URL. */
fun newSlack(url: String): Notifier = SlackNotifier(url)

/** Renders a Message as Slack mrkdwn. */
internal fun slackText(m: Message): String {
    val b = StringBuilder()
    if (m.title.isNotEmpty()) {
        b.append("*").append(m.title).append("*")
    }
    if (m.text.isNotEmpty()) {
        if (b.isNotEmpty()) b.append("\n")
        b.append(m.text)
    }
    if (m.link.isNotEmpty()) {
        if (b.isNotEmpty()) b.append("\n")
        b.append("<").append(m.link).append(">")
    }
    return b.toString()
}
